pkg/plugin/validator.go

```diff
@@ -0,0 +1,182 @@
+	pk, err := readPrivateKey(cfg.GithubAppPrivateKeyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read private key: %w", err)
+	}
+	v := Validator{
+		cfg:        cfg,
+		decodedPem: pk,
+		client:     github.NewClient(nil),
+	}
+	for _, opt := range opts {
+		opt(&v)
+	}
+	return &v, nil
+}
+
+// MatchIssue parses issue info from provided issueURL and validate if the issue is valid.
+func (v *Validator) MatchIssue(ctx context.Context, issueURL string, opts ...githubapp.ConfigOption) error {
+	info, err := parseIssueInfoFromURL(issueURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse issueURL: %w", err)
+	}
+
+	t, err := v.getAccessToken(ctx, v.cfg, v.decodedPem, info.RepoName, opts)
+	if err != nil {
+		return fmt.Errorf("failed to get access token: %w", err)
+	}
+	v.client = v.client.WithAuthToken(t)
+
+	return v.validateIssue(ctx, info)
+}
+
+// validateIssue verifies if the issue exists and the issue is open.
+func (v *Validator) validateIssue(ctx context.Context, pi *pluginGithubIssue) error {
+	issue, err := v.getGithubIssue(ctx, pi)
+	if err != nil {
+		return fmt.Errorf("failed to get issue info: %w", err)
+	}
+	if s := issue.GetState(); s != "open" {
+		return fmt.Errorf("issue is in state: %s, please make sure to use an open issue", s)
+	}
+	return nil
+}
+
+// getGithubIssue gets the provided issue's info from github api.
+func (v *Validator) getGithubIssue(ctx context.Context, pi *pluginGithubIssue) (*github.Issue, error) {
+	if v.fakeGithubIssue != nil {
+		return v.fakeGithubIssue, nil
+	}
+	issue, _, err := v.client.Issues.Get(ctx, pi.Owner, pi.RepoName, pi.IssueNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get issue: %w", err)
+	}
+	return issue, nil
+}
+
+// getAccessToken gets an access token with issue read permission to the repo
+// which contains the issue.
+func (v *Validator) getAccessToken(ctx context.Context, cfg *PluginConfig, pk *rsa.PrivateKey, repoName string, opts []githubapp.ConfigOption) (string, error) {
+	opts = append(opts, githubapp.WithJWTTokenCaching(1*time.Minute))
+
+	ghCfg := githubapp.NewConfig(cfg.GithubAppID, cfg.GithubAppInstallationID, pk, opts...)
+	githubApp := githubapp.New(ghCfg)
+
+	tr := githubapp.TokenRequest{
+		Repositories: []string{repoName},
+		Permissions: map[string]string{
+			"issues": "read",
+		},
+	}
+	var tokenResp ExchangeResponse
+	resp, err := githubApp.AccessToken(ctx, &tr)
+	if err != nil {
+		return "", fmt.Errorf("failed to get access token: %w", err)
+	}
+
+	if err := json.Unmarshal([]byte(resp), &tokenResp); err != nil {
+		return "", fmt.Errorf("error unmarshal resp: %w", err)
+	}
+	return tokenResp.AccessToken, nil
+}
+
+// parseGithubIssue parses issue info from Issue URL.
+func parseIssueInfoFromURL(issueURL string) (*pluginGithubIssue, error) {
+	u, err := url.Parse(issueURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse provided issue url: %w", err)
```

Comment:
No action required at this point. I believe you will soon need to differentiate "validation error" vs. "internal error"

Reply:
Created issues#3 to track this.